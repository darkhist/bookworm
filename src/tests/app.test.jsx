import React from 'react';
import { shallow } from 'enzyme';
import mock from 'axios';

import App from '../components/index';
import Err from '../components/err/err';

const valid = 'https://www.googleapis.com/books/v1/volumes?q=flowers';
const badReq = 'https://www.googleapis.com/books/v1/volumes?q=%%%%';
const noResults = 'https://www.googleapis.com/books/v1/volumes?q=really_long_strange_query';
const invalidURL = 'https://www.googleapis.com/books/v9999';
const getBooks = async url => mock.get(`${url}`);

describe('App', () => {
  it('should render without crashing', () => {
    const app = shallow(<App />);
    expect(app).toMatchSnapshot();
  });

  it('should have the correct init state', () => {
    const app = shallow(<App />);
    expect(app.state('results')).toEqual([]);
    expect(app.state('valid')).toBeTruthy();
    expect(app.state('error')).toBeFalsy();
  });

  it('should call getBooks on componentDidMount', () => {
    const spy = jest.spyOn(App.prototype, 'getBooks');
    shallow(<App />);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith('cats');
  });

  it('should have the correct state on valid queries', async () => {
    const app = shallow(<App />);
    const instance = app.instance();
    const spy = jest.spyOn(instance, 'getBooks');
    instance.getBooks(valid);
    expect(spy).toHaveBeenCalledTimes(2);

    const results = await getBooks(valid);
    expect(app.state('results')).toEqual(results.data.items);
    expect(app.state('valid')).toBeTruthy();
    expect(app.state('error')).toBeFalsy();
  });

  it('should have the correct state on invalid queries', () => {
    const app = shallow(<App />);
    const instance = app.instance();
    const spy = jest.spyOn(instance, 'getBooks');
    instance.getBooks(badReq);
    expect(spy).toHaveBeenCalledTimes(2);
    expect(app.state('valid')).toBeFalsy();
    expect(app).toMatchSnapshot();
  });

  it('should have the correct state on API call failure', async () => {
    const app = shallow(<App />);
    const instance = app.instance();
    const spy = jest.spyOn(instance, 'setState');
    await instance.getBooks(invalidURL);
    expect(spy).toHaveBeenCalled();
    expect(app.find(Err)).toBeTruthy();
  });

  it('should handle state changes properly', async () => {
    const app = shallow(<App />);
    const instance = app.instance();
    const spy = jest.spyOn(instance, 'getBooks');
    await instance.getBooks(valid);
    expect(spy).toHaveBeenCalledTimes(2);
    expect(app.state('results')).toHaveLength(1);
    expect(app.state('valid')).toBeTruthy();
    expect(app.state('error')).toBeFalsy();
    await instance.getBooks(badReq);
    expect(spy).toHaveBeenCalledTimes(3);
    expect(app.state('results')).toHaveLength(1);
    expect(app.state('valid')).toBeFalsy();
    expect(app.state('error')).toBeFalsy();
    expect(app).toMatchSnapshot();
  });

  it('should render errors accordingly', () => {
    const app = shallow(<App />);
    app.setState({ error: true });
    expect(app.find(Err)).toBeTruthy();
    expect(app).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});


describe('getBooks()', () => {
  it('should query Google Books API & return correct info', async () => {
    const query = await getBooks(valid);

    expect(query.data).toBeDefined();

    const book = query.data.items[0].volumeInfo;

    const info = {
      title: book.title,
      authors: book.authors,
      publisher: book.publisher,
      thumbnail: book.imageLinks.smallThumbnail,
      link: book.infoLink
    };

    expect(info.title).toEqual('Flowers');
    expect(info.authors).toEqual(['Gail Saunders Smith']);
    expect(info.publisher).toEqual('Capstone');
    expect(info.thumbnail).toEqual('http://books.google.com/books/content?id=ogs_KDUQLSsC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api');
    expect(info.link).toEqual('http://books.google.com/books?id=ogs_KDUQLSsC&dq=flowers&hl=&source=gbs_api');
  });

  it('should detect if no results are found', async () => {
    try {
      const results = await getBooks(noResults);
      expect(results.data.items.length).toBe(0);
    } catch (error) {
      expect(error.status).toBe(200);
      expect(error.data).toEqual({
        kind: 'books#volumes',
        totalItems: 0
      });
    }
  });

  it('should detect if invalid terms are used', async () => {
    try {
      await getBooks(badReq);
    } catch (error) {
      expect(error.status).toBe(400);
      expect(error.data).toEqual('Bad Request');
    }
  });

  it('should detect API request failures', async () => {
    try {
      const results = await getBooks(invalidURL);
      expect(results.data.items.length).toBe(0);
      const app = shallow(<App />);
      expect(app).toMatchSnapshot();
    } catch (error) {
      expect(error.status).toBe(404);
      expect(error.data).toEqual('Not Found');
    }
  });
});
