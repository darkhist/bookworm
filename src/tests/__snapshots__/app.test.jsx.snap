// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should handle state changes properly 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <Header />
  <Search
    onChange={[Function]}
    valid={false}
  />
  <Results
    results={
      Array [
        Object {
          "accessInfo": Object {
            "accessViewStatus": "SAMPLE",
            "country": "US",
            "embeddable": true,
            "epub": Object {
              "isAvailable": false,
            },
            "pdf": Object {
              "isAvailable": false,
            },
            "publicDomain": false,
            "quoteSharingAllowed": false,
            "textToSpeechPermission": "ALLOWED",
            "viewability": "PARTIAL",
            "webReaderLink": "http://play.google.com/books/reader?id=ogs_KDUQLSsC&hl=&printsec=frontcover&source=gbs_api",
          },
          "etag": "CfGDkHrf3vo",
          "id": "ogs_KDUQLSsC",
          "kind": "books#volume",
          "saleInfo": Object {
            "country": "US",
            "isEbook": false,
            "saleability": "NOT_FOR_SALE",
          },
          "searchInfo": Object {
            "textSnippet": "Simple text and photographs depict the parts of flowers and their pollination.",
          },
          "selfLink": "https://www.googleapis.com/books/v1/volumes/ogs_KDUQLSsC",
          "volumeInfo": Object {
            "allowAnonLogging": false,
            "authors": Array [
              "Gail Saunders Smith",
            ],
            "canonicalVolumeLink": "https://books.google.com/books/about/Flowers.html?hl=&id=ogs_KDUQLSsC",
            "categories": Array [
              "Juvenile Nonfiction",
            ],
            "contentVersion": "0.1.1.0.preview.1",
            "description": "Simple text and photographs depict the parts of flowers and their pollination.",
            "imageLinks": Object {
              "smallThumbnail": "http://books.google.com/books/content?id=ogs_KDUQLSsC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
              "thumbnail": "http://books.google.com/books/content?id=ogs_KDUQLSsC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
            },
            "industryIdentifiers": Array [
              Object {
                "identifier": "0736848649",
                "type": "ISBN_10",
              },
              Object {
                "identifier": "9780736848640",
                "type": "ISBN_13",
              },
            ],
            "infoLink": "http://books.google.com/books?id=ogs_KDUQLSsC&dq=flowers&hl=&source=gbs_api",
            "language": "en",
            "maturityRating": "NOT_MATURE",
            "pageCount": 24,
            "previewLink": "http://books.google.com/books?id=ogs_KDUQLSsC&printsec=frontcover&dq=flowers&hl=&cd=1&source=gbs_api",
            "printType": "BOOK",
            "publishedDate": "2004-09-01",
            "publisher": "Capstone",
            "readingModes": Object {
              "image": true,
              "text": false,
            },
            "title": "Flowers",
          },
        },
      ]
    }
  />
</Fragment>
`;

exports[`App should have the correct state on invalid queries 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <Header />
  <Search
    onChange={[Function]}
    valid={false}
  />
  <Results
    results={Array []}
  />
</Fragment>
`;

exports[`App should render errors accordingly 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <Header />
  <Search
    onChange={[Function]}
    valid={true}
  />
  <Err
    msg="Oops. We couldn't reach the Google Books API"
  />
</Fragment>
`;

exports[`App should render without crashing 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <Header />
  <Search
    onChange={[Function]}
    valid={true}
  />
  <Results
    results={Array []}
  />
</Fragment>
`;
